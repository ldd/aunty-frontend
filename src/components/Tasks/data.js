const generateFakeData = () =>
  Array(31)
    .fill(true)
    .map(_ => Math.random() < 0.5);

export const prepareData = (data = []) =>
  data.reduce(
    (acc, n, i) =>
      i % 7 === 0
        ? [...acc, [n]]
        : [...acc.slice(0, acc.length - 1), acc[acc.length - 1].concat(n)],
    []
  );

const TASKS_URL = "http://localhost:4000/v1/tasks";

export const fetchData = async (
  setter = () => {},
  failer = () => {},
  url = TASKS_URL,
  options = {}
) => {
  const rawData = await fetch(url, {
    credentials: "include",
    ...options
  });
  if (rawData.ok) {
    const data = await rawData.json();
    setter(data);
  } else {
    failer();
  }
};

export const deleteTask = async (id = -1, setter, failer) =>
  fetchData(setter, failer, `${TASKS_URL}/${id}`, { method: "DELETE" });

export const fakeTasks = [
  {
    label: "Gym",
    description: "Go to the Gym",
    data: generateFakeData(),
    id: "id#Gym"
  },
  {
    label: "Read",
    description: "Read 10 pages",
    data: generateFakeData(),
    id: "id#Read"
  }
];
