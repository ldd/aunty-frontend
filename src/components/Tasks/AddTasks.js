import React, { useEffect, useState, useCallback } from "react";
import Heading from "../Heading";
import { addTask } from "./data";

const Field = ({ label, value, setValue, hint }) => {
  const onChange = e => setValue(e.target.value);
  return (
    <div className="field" style={{ textAlign: "left" }}>
      <label className="label">{label}</label>
      <div className="control has-icons-left has-icons-right">
        <input
          className="input is-success"
          type="text"
          value={value}
          onChange={onChange}
        />
        <span className="icon is-small is-left">
          <i className="fas fa-user" />
        </span>
        <span className="icon is-small is-right">
          <i className="fas fa-check" />
        </span>
      </div>
      <p className="help is-dark">{hint}</p>
    </div>
  );
};

const Control = ({ onSubmit, onCancel }) => (
  <div className="field is-grouped is-grouped-right">
    <div className="control">
      <div onClick={onCancel} className="button is-light">
        Cancel
      </div>
    </div>
    <div className="control">
      <div onClick={onSubmit} className="button is-primary">
        Submit
      </div>
    </div>
  </div>
);
const Form = ({ hideModal, onTaskAdded }) => {
  // field data
  const [label, setLabel] = useState("");
  const [description, setDescription] = useState("");
  const [frequency, setFrequency] = useState("");
  // for submitting the form
  const [payload, setPayload] = useState(null);
  const onSubmit = () => {
    setPayload({
      task_label: label,
      task_description: description,
      task_frequency: frequency
    });
  };
  useEffect(() => {
    if (payload !== null) {
      addTask(
        JSON.stringify(payload),
        (...args) => {
          hideModal();
          onTaskAdded(...args);
        },
        hideModal
      );
    }
  }, [payload, hideModal, onTaskAdded]);

  return (
    <div className="box">
      <Field
        label="Label"
        value={label}
        hint="example: read"
        setValue={setLabel}
      />
      <Field
        label="Description"
        value={description}
        hint="example: read at least 10 pages"
        setValue={setDescription}
      />
      <Field
        label="Frequency"
        value={frequency}
        hint="examples: monday, wednesday, friday | daily | weekends"
        setValue={setFrequency}
      />
      <Control onSubmit={onSubmit} onCancel={hideModal} />
    </div>
  );
};
const AddTasks = ({ onTaskAdded }) => {
  const [modal, setModal] = useState(false);
  const showModal = () => setModal(true);
  const hideModal = useCallback(() => setModal(false), []);
  return (
    <>
      <div
        onClick={showModal}
        className="button is-large is-success is-inverted"
        style={{ fontSize: "2.5rem" }}
      >
        +
      </div>
      <div className={`modal ${modal ? "is-active" : ""}`}>
        <div className="modal-background" />
        <div className="modal-content">
          <Heading title="Add Task" />
          <Form hideModal={hideModal} onTaskAdded={onTaskAdded} />
        </div>
        <button
          onClick={hideModal}
          className="modal-close is-large"
          aria-label="close"
        />
      </div>
    </>
  );
};

export default AddTasks;
