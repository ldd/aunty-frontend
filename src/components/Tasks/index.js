import React, { useState, useEffect } from "react";
import Heading from "../Heading";
import AddTasks from "./AddTasks";
import { prepareData, fetchData, deleteTask } from "./data";
import { redirectToLogin } from "../Login";

export const Entry = ({ done = false }) => (
  <div
    style={{
      width: 12,
      margin: 1.5,
      height: 12,
      backgroundColor: done ? "rgba(0,200,0,0.8)" : "rgba(0,200,0,0.2)"
    }}
  />
);

export const Task = ({ label, description, data = [], deleteTask }) => (
  <>
    <div style={{ position: "relative" }}>
      <div
        className="delete"
        onClick={deleteTask}
        style={{ position: "absolute", right: 0, top: -10 }}
      />
      <Heading title={label} subtitle={description} />
    </div>
    <div className="box" style={{ display: "flex", flexDirection: "column" }}>
      {prepareData(data).map((weekData, j) => (
        <div key={`week-${j}`} style={{ display: "flex" }}>
          {weekData.map((el, day) => (
            <Entry key={`week-${j}-day-${day}`} done={el} />
          ))}
        </div>
      ))}
    </div>
  </>
);

const Tasks = ({ initialData = [] }) => {
  const [data, setData] = useState(initialData);
  const [deleteId, setDeleteId] = useState(null);
  const [addedId, setAddedId] = useState(null);

  useEffect(() => {
    fetchData(setData, redirectToLogin);
  }, [deleteId, addedId]);
  useEffect(() => {
    if (typeof deleteId === "string") {
      deleteTask(deleteId, setDeleteId);
    }
  }, [deleteId]);
  return (
    <div className="hero-body">
      <div className="container has-text-centered">
        <div className="columns is-multiline is-vcentered">
          {data.map(({ id, ...rest }) => (
            <div key={id} className="column is-narrow">
              <Task key={id} {...rest} deleteTask={() => setDeleteId(id)} />
            </div>
          ))}
          <div className="column is-narrow">
            <AddTasks onTaskAdded={setAddedId} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default Tasks;
