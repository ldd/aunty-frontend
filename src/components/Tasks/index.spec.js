import React from "react";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { shallow, mount } from "enzyme";
import Tasks, { Task, Entry } from ".";
import * as data from "./data";

const { generateFakeData, fakeTasks } = data;
global.Math.random = () => 0.5;

describe("Heading", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fetch.resetMocks();
    const mockedAddTask = jest.fn().mockImplementation(() => fakeTasks);
    jest.spyOn(data, "fetchData").mockImplementation(mockedAddTask);
  });

  it("renders snapshot [tasks]", () => {
    const component = renderer.create(<Tasks />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders snapshot [task]", () => {
    const component = renderer.create(<Task />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders snapshot [task with data]", () => {
    const component = renderer.create(<Task data={generateFakeData()} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders snapshot [entry]", () => {
    const component = renderer.create(<Entry />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders mounted tasks", async () => {
    fetch.mockResponse(JSON.stringify(fakeTasks), { ok: true });

    const wrapper = mount(<Tasks initialData={fakeTasks} />);
    const getTask = () => wrapper.find(Task).first();
    await act(async () => await getTask().prop("deleteTask")());

    const label = getTask().prop("label");
    const [url, options] = fetch.mock.calls[0];
    expect(url).toContain(label);
    expect(options).toHaveProperty("method", "DELETE");
  });
});
