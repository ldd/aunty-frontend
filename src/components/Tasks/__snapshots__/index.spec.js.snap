// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading renders snapshot [entry] 1`] = `
<div
  style={
    Object {
      "backgroundColor": "rgba(0,200,0,0.2)",
      "height": 12,
      "margin": 1.5,
      "width": 12,
    }
  }
/>
`;

exports[`Heading renders snapshot [task with data] 1`] = `
Array [
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="delete"
      style={
        Object {
          "position": "absolute",
          "right": 0,
          "top": -10,
        }
      }
    />
  </div>,
  <div
    className="box"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
      <div
        style={
          Object {
            "backgroundColor": "rgba(0,200,0,0.2)",
            "height": 12,
            "margin": 1.5,
            "width": 12,
          }
        }
      />
    </div>
  </div>,
]
`;

exports[`Heading renders snapshot [task] 1`] = `
Array [
  <div
    style={
      Object {
        "position": "relative",
      }
    }
  >
    <div
      className="delete"
      style={
        Object {
          "position": "absolute",
          "right": 0,
          "top": -10,
        }
      }
    />
  </div>,
  <div
    className="box"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  />,
]
`;

exports[`Heading renders snapshot [task] 2`] = `
<div
  style={
    Object {
      "backgroundColor": "rgba(0,200,0,0.2)",
      "height": 12,
      "margin": 1.5,
      "width": 12,
    }
  }
/>
`;

exports[`Heading renders snapshot [tasks] 1`] = `
<div
  className="hero-body"
>
  <div
    className="container has-text-centered"
  >
    <div
      className="columns is-multiline is-vcentered"
    >
      <div
        className="column is-narrow"
      >
        <div
          className="button is-large is-success is-inverted"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "2.5rem",
            }
          }
        >
          +
        </div>
        <div
          className="modal "
        >
          <div
            className="modal-background"
          />
          <div
            className="modal-content"
          >
            <h1
              className="title"
            >
              Add Task
            </h1>
            <div
              className="box"
            >
              <div
                className="field"
                style={
                  Object {
                    "textAlign": "left",
                  }
                }
              >
                <label
                  className="label"
                >
                  Label
                </label>
                <div
                  className="control has-icons-left has-icons-right"
                >
                  <input
                    className="input is-success"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                  <span
                    className="icon is-small is-left"
                  >
                    <i
                      className="fas fa-user"
                    />
                  </span>
                  <span
                    className="icon is-small is-right"
                  >
                    <i
                      className="fas fa-check"
                    />
                  </span>
                </div>
                <p
                  className="help is-dark"
                >
                  example: read
                </p>
              </div>
              <div
                className="field"
                style={
                  Object {
                    "textAlign": "left",
                  }
                }
              >
                <label
                  className="label"
                >
                  Description
                </label>
                <div
                  className="control has-icons-left has-icons-right"
                >
                  <input
                    className="input is-success"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                  <span
                    className="icon is-small is-left"
                  >
                    <i
                      className="fas fa-user"
                    />
                  </span>
                  <span
                    className="icon is-small is-right"
                  >
                    <i
                      className="fas fa-check"
                    />
                  </span>
                </div>
                <p
                  className="help is-dark"
                >
                  example: read at least 10 pages
                </p>
              </div>
              <div
                className="field"
                style={
                  Object {
                    "textAlign": "left",
                  }
                }
              >
                <label
                  className="label"
                >
                  Frequency
                </label>
                <div
                  className="control has-icons-left has-icons-right"
                >
                  <input
                    className="input is-success"
                    onChange={[Function]}
                    type="text"
                    value=""
                  />
                  <span
                    className="icon is-small is-left"
                  >
                    <i
                      className="fas fa-user"
                    />
                  </span>
                  <span
                    className="icon is-small is-right"
                  >
                    <i
                      className="fas fa-check"
                    />
                  </span>
                </div>
                <p
                  className="help is-dark"
                >
                  examples: monday, wednesday, friday | daily | weekends
                </p>
              </div>
              <div
                className="field is-grouped is-grouped-right"
              >
                <div
                  className="control"
                >
                  <div
                    className="button is-light"
                    onClick={[Function]}
                  >
                    Cancel
                  </div>
                </div>
                <div
                  className="control"
                >
                  <div
                    className="button is-primary"
                    onClick={[Function]}
                  >
                    Submit
                  </div>
                </div>
              </div>
            </div>
          </div>
          <button
            aria-label="close"
            className="modal-close is-large"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
